iwyu_ast_util.cc

```diff
@@ -546,32 +560,32 @@ class TypeEnumerator : public RecursiveASTVisitor<TypeEnumerator> {
 
   // --- Methods on RecursiveASTVisitor
   bool TraverseType(QualType type) {
-    TraverseArgumentsOfSugaredTemplates(type);
+    TraverseTypeHelper(type);
     return Base::TraverseType(type);
   }
 
   bool TraverseTypeLoc(TypeLoc type_loc) {
-    TraverseArgumentsOfSugaredTemplates(type_loc.getType());
+    TraverseTypeHelper(type_loc.getType());
     return Base::TraverseTypeLoc(type_loc);
   }
 
-  bool VisitType(Type* type) {
-    seen_types_.insert(type);
-    return true;
-  }
-
  private:
-  // Clang doesn't traverse underlying type for most of the sugar types.
-  // If a TemplateSpecializationType occurs in a sugaring chain, traverse its
-  // arguments explicitly, otherwise they aren't put into resugar_map.
-  void TraverseArgumentsOfSugaredTemplates(QualType type) {
-    if (type.isNull())
+  void TraverseTypeHelper(QualType qual_type) {
+    if (qual_type.isNull())
       return;
 
-    if (const auto* template_spec = type->getAs<TemplateSpecializationType>()) {
-      for (const TemplateArgument& arg : template_spec->template_arguments())
-        TraverseTemplateArgument(arg);
-    }
+    const Type* const type = qual_type.getTypePtr();
+    const Type* const desugared_until_alias_or_tpl = Desugar(type);
+    seen_types_.insert(desugared_until_alias_or_tpl);
+
+    const Type* const desugared =
+        DesugarAliasTypes(desugared_until_alias_or_tpl);
+    // A desugared type shouldn't be inserted into resugar_map along with its
+    // sugared version as a value instead of a key, hence it is important to
+    // call the base class implementation of TraverseType and not the own.
+    // TODO(bolshakov): check whether typedef components are "provided".
```

Comment:
It would be nice if we didn't have to do that in here -- most of that logic lives in iwyu.cc where we have more of a high-level view.

Reply:
I don't see at now any convenient way to do this except moving `TypeEnumerator` in iwyu.cc along with all the dependent functions...

Reply:
I suppose we can move the mechanics for checking back in here when the time comes. Hopefully.